proxy_agent/src/data_vessel.rs

```diff
@@ -0,0 +1,93 @@
+// Copyright (c) Microsoft Corporation
+// SPDX-License-Identifier: MIT
+
+use crate::key_keeper::key::Key;
+use std::sync::mpsc::Sender;
+
+pub enum DataAction {
+    Stop,
+    KeyKeeperSetKey {
+        key: Key,
+        response: std::sync::mpsc::Sender<bool>,
+    },
+    KeyKeeperGetKeyValue {
+        response: std::sync::mpsc::Sender<Key>,
+    },
+}
+
+#[derive(Clone, Debug)]
+pub struct DataVessel {
+    sender: Sender<DataAction>,
+}
+impl DataVessel {
+    pub fn start_new_async() -> Self {
+        let (sender, receiver) = std::sync::mpsc::channel::<DataAction>();
+
+        std::thread::spawn(move || {
+            // chached data are defined here
+            let mut cached_key: Key = Key::empty();
+
+            while let Ok(action) = receiver.recv() {
+                match action {
+                    DataAction::Stop => {
+                        break;
+                    }
+                    DataAction::KeyKeeperSetKey { key, response } => {
+                        // Set the key from the key keeper
+                        cached_key = key.clone();
+                        _ = response.send(true);
+                    }
+                    DataAction::KeyKeeperGetKeyValue { response } => {
+                        // Get the key from the key keeper
+                        _ = response.send(cached_key.clone());
+                    }
+                }
+            }
+        });
+
+        DataVessel { sender }
+    }
+
+    pub fn stop(&self) {
```

Comment:
It might be better to consume `self` since there's no way to use this object after calling `stop`:
```suggestion
    pub fn stop(self) {
```

Obviously since this structure is clone-able this doesn't stop other users from trying to use it after its stopped, but it is a good signal that's the intention.

Reply:
good suggestion.

Reply:
This doesn't seem to have been actually resolved.

Reply:
good catch, seems like the change got reverted from my local. fixed it.